### ISSUE TITLE
4425 - Greater Than

### ISSUE THREAD

**Issue by ohk990102** [type-challenges/type-challenges#5077]
```ts
type ArrayWithLength<T extends number, U extends any[] = []> = U['length'] extends T ? U : ArrayWithLength<T, [true, ...U]>;
type GreaterThan<T extends number, U extends number> = ArrayWithLength<U> extends [...ArrayWithLength<T>, ...infer _] ? false : true;
```

**Comment by qianghe** [comment]
resolved big number case:
```typescript
  type ParseInt<T> = T extends `${infer X extends number}` ? X :  never

  type RemoveLeadingZeros<T extends string> = T extends '0' ? T : (
    T extends `${0}${infer Rest}` ? RemoveLeadingZeros<Rest> : T
  )
  
  type InnerMinusOne<T extends string> = T extends `${infer X extends number}${infer Y}` ? (
    X extends 0 ? `9${InnerMinusOne<Y>}` : `${[-1, 0, 1, 2, 3, 4, 5, 6, 7, 8][X]}${Y}`
  ) : ''
  
  type Reverse<T extends string> = T extends `${infer X}${infer Y}` ? `${Reverse<Y>}${X}` : ''
  
  type MinusOne<T extends number> = ParseInt<RemoveLeadingZeros<Reverse<InnerMinusOne<Reverse<`${T}`>>>>>
  
  type InnerGreaterThan<T extends number, U extends number> = T extends U ? true : (
    T extends 0 ? false : InnerGreaterThan<MinusOne<T>, U>
  )

  type GreaterThan<T extends number, U extends number> = T extends U ? false : (
    U extends 0 ? true : InnerGreaterThan<T, U>
  )
```

**Comment by sv-98-maxin** [comment]
resolve big number case
```ts
type StringToArray<T extends string> = T extends `${infer L extends number}${infer Rest}` ? [L, ...StringToArray<Rest>] : []
type CreateArrayByNumber<T extends number, R extends any[] = []> = R['length'] extends T ? R : CreateArrayByNumber<T, [...R, undefined]>
type CompareOneDigit<T extends number, U extends number> = T extends U ? false :  CreateArrayByNumber<T> extends [...CreateArrayByNumber<U>, ...infer _] ? true : false
type CompareAllNumber<T extends any[], U extends any[], CompareCount extends any[] = []> = T['length'] extends U['length'] ? CompareCount['length'] extends T['length'] ? false : CompareOneDigit<T[CompareCount['length']], U[CompareCount['length']]> extends true ? true : CompareAllNumber<T, U, [...CompareCount, 1]> : T[U['length']] extends undefined ? false : true
type GreaterThan<T extends number, U extends number> = CompareAllNumber<StringToArray<`${T}`>, StringToArray<`${U}`>>
```

**Comment by Eiriksm0raas** [comment]
Probably not optimal, but a bit simpler (for the big number)
```ts
type ArrayOfLengt<N extends number, A extends any[] = []> =
  A["length"] extends N ? A : ArrayOfLengt<N, [any, ...A]>

type DGreaterThan<
  T extends number, 
  U extends number,
  TA extends any[] = ArrayOfLengt<T>,
  UA extends any[] = ArrayOfLengt<U>
> = 
  TA extends [any, ...infer TR] ?
    UA extends [any, ...infer UR] ?
      DGreaterThan<number, number, TR, UR> :
    true :
  false

type Length<T extends string, A extends string[] = []> =
  T extends `${infer F extends string}${infer R extends string}` 
    ? Length<R, [...A, F]>
    : A["length"]

type GreaterThan<
  T extends number, 
  U extends number,
  TS extends string = `${T}`,
  US extends string = `${U}`
> = 
  DGreaterThan<Length<TS>, Length<US>> extends true ? true :
  DGreaterThan<Length<US>, Length<TS>> extends true ? false :
  TS extends `${infer TF extends number}${infer TR extends string}` ?
    US extends `${infer UF extends number}${infer UR extends string}` ?
      DGreaterThan<TF, UF> extends true ? true :
      DGreaterThan<UF, TF> extends true ? false :
      GreaterThan<number, number, TR, UR> :
    false :
  false
```

**Comment by likepeach789** [comment]
> ArrayWithLength extends [...ArrayWithLength, ...infer _] ? false : true;
> 
> how it works?

ArrayWithLength<U> = [true, true...., true] // The array length is U
ArrayWithLength<T> = [true, true...., true] // The array length is T

ArrayWithLength<U> extends [...ArrayWithLength<T>, ...infer _] ? false : true;
 = [true, true...., true] extends [true, true...., true, infer _] ? false : true

If U <= T, then  [true, true...., true] extends [true, true...., true, infer _]  returns false otherwise is true. extends keyword means the left one could be assigned to the right.

**Comment by mg901** [comment]
type ToDigits<S extends string> = S extends `${infer D}${infer Rest}`
  ? [D, ...ToDigits<Rest>]
  : [];

type DigitGreaterThan<A extends string, B extends string> = 
  A extends '0' ? (B extends '0' ? false : false) :
  A extends '1' ? (B extends '0' ? true : B extends '1' ? false : false) :
  A extends '2' ? (B extends '0' | '1' ? true : B extends '2' ? false : false) :
  A extends '3' ? (B extends '0' | '1' | '2' ? true : B extends '3' ? false : false) :
  A extends '4' ? (B extends '0' | '1' | '2' | '3' ? true : B extends '4' ? false : false) :
  A extends '5' ? (B extends '0' | '1' | '2' | '3' | '4' ? true : B extends '5' ? false : false) :
  A extends '6' ? (B extends '0' | '1' | '2' | '3' | '4' | '5' ? true : B extends '6' ? false : false) :
  A extends '7' ? (B extends '0' | '1' | '2' | '3' | '4' | '5' | '6' ? true : B extends '7' ? false : false) :
  A extends '8' ? (B extends '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' ? true : B extends '8' ? false : false) :
  A extends '9' ? (B extends '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' ? true : B extends '9' ? false : false) :
  never;

type GreaterThanHelper<
  A extends string[],
  B extends string[]
> = A extends [infer ADigit extends string, ...infer ARest extends string[]]
  ? B extends [infer BDigit extends string, ...infer BRest extends string[]]
    ? DigitGreaterThan<ADigit, BDigit> extends true
      ? true
      : DigitGreaterThan<ADigit, BDigit> extends false
      ? GreaterThanHelper<ARest, BRest>
      : false
    : true // A is longer than B
  : false; // B is longer than A

type GreaterThan<A extends number, B extends number> = GreaterThanHelper<
  ToDigits<`${A}`>,
  ToDigits<`${B}`>
>;

**Comment by TheAcebyte** [comment]
> ArrayWithLength extends [...ArrayWithLength, ...infer _] ? false : true;
> 
> how it works?

If there's remaining space in the array to infer _, that means our first array was larger than the second one. Therefore, U was greater than T.
Say for example: T = 2 and U = 1, ArrayWithLength<> just returns a tuple type filled with "true" whose length is the passed in generic parameter, then, we get ArrayWithLength<T> = [true, true] and ArrayWithLength<U> = [true]. This means ArrayWithLength<T> can extend [ArrayWithLength<U>, ...infer _] because it can infer _ as the missing [true] value, which is why T is greater than U.
In other words, we're trying to compare T and U by comparing tuples with their respective lengths.

**Comment by zhangmingdi** [comment]
type GreaterThanImpl<T extends number,K extends number,Arr extends unknown[] = [],Bool extends number = 0> = Arr['length'] extends T 
?  Bool : (
    GreaterThanImpl<T,K,[...Arr,1],
        Bool extends 0 ? ([...Arr,1]['length'] extends K ? 1 :Bool):( Bool extends 1 ? 2 :Bool)
    >
)
type GreaterThan<T extends number,K extends number,> = GreaterThanImpl<T,K> extends 0 | 1 ? false :true

**Comment by Withered-Flower-0422** [comment]
Resolve negative number, decimal number, super big number (e.g. `1e25`), super small number (e.g. `1e-10`) cases:
[Compare](https://github.com/type-challenges/type-challenges/issues/37189#issuecomment-3214329176)

**Comment by Barrenboat** [comment]
Genius code. It could be a ​​tool for resolving big number​.

**Comment by q2177329q** [comment]
```

type DigitMap = {
  "0": 0;
  "1": 1;
  "2": 2;
  "3": 3;
  "4": 4;
  "5": 5;
  "6": 6;
  "7": 7;
  "8": 8;
  "9": 9;
};
type NumberToByteArray<
  T extends number | string,
  A extends number[] = []
> = `${T}` extends `${infer L extends keyof DigitMap}${infer R}`
  ? NumberToByteArray<R, [...A, DigitMap[L]]>
  : A;

type NumberToArray<T extends number> = T extends 0
  ? []
  : T extends 1
  ? [1]
  : [1, ...NumberToArray<[0, 0, 1, 2, 3, 4, 5, 6, 7, 8][T]>];
type GreaterThanSingle<
  T extends number,
  U extends number
> = NumberToArray<U> extends [...NumberToArray<T>, ...infer _] ? false : true;

type NumberToLengthArray<
  T extends number | string,
  A extends any[] = []
> = `${T}` extends `${infer L}${infer R}`
  ? L extends ""
    ? A
    : NumberToLengthArray<R, [...A, 1]>
  : A;

type GreaterThanByLength<
  T extends number,
  U extends number
> = NumberToLengthArray<U> extends [...NumberToLengthArray<T>, ...infer _]
  ? false
  : true;

type GreaterThanByByte<T extends number[], U extends number[]> = T extends [
  infer L1 extends number,
  ...infer R1 extends number[]
]
  ? U extends [infer L2 extends number, ...infer R2 extends number[]]
    ? true extends GreaterThanSingle<L1, L2>
      ? true
      : GreaterThanByByte<R1, R2>
    : false
  : false;

type GreaterThan<T extends number, U extends number> = U extends T
  ? false
  : true extends GreaterThanByLength<T, U>
  ? true
  : GreaterThanByByte<NumberToByteArray<T>, NumberToByteArray<U>>;
```
- 先比较是否全等
- 然后比较长度，更长的更大
- 最后剩下两个数位数一样，按位比较

**Comment by XinXiaoIsMe** [comment]
```ts
type NumberToTuple<N extends number, R extends any[] = []> = N extends R['length'] ? R : NumberToTuple<N, [1, ...R]>;

type CompareSmallNumber<N1 extends number, N2 extends number> = NumberToTuple<N1> extends [...NumberToTuple<N2>, ...infer R]
  ? R[0] extends 1
    ? 'greater'
    : 'equal'
  : 'less'

type StrToNum<S extends string> = S extends `${infer N extends number}` ? N : never

type StrToTuple<S extends string> = S extends `${infer F}${infer R}` ? [StrToNum<F>, ...StrToTuple<R>] : []

type CompareTuple<T1 extends number[], T2 extends number[], CurrentIndex extends any[] = []> = CurrentIndex['length'] extends T1['length']
  ? false
  : CompareSmallNumber<T1[CurrentIndex['length']], T2[CurrentIndex['length']]> extends 'greater'
    ? true
    : CompareSmallNumber<T1[CurrentIndex['length']], T2[CurrentIndex['length']]> extends 'less'
      ? false
      : CompareTuple<T1, T2, [1, ...CurrentIndex]>;

type GreaterThan<
  T extends number, 
  U extends number,
  TLength extends number = StrToTuple<`${T}`>['length'],
  ULength extends number = StrToTuple<`${U}`>['length'],
> = CompareSmallNumber<TLength, ULength> extends 'less'
  ? false
  : CompareSmallNumber<TLength, ULength> extends 'greater'
    ? true
    : TLength extends ULength
      ? CompareTuple<StrToTuple<`${T}`>, StrToTuple<`${U}`>>
      : false
```

首先判断长度，长度不一致很容易判断。长度一致时，按位比较。
